### DISCUSSION TITLE
Keeping things safe with supabase and cloud functions.

### DISCUSSION THREAD

**Discussion by koaning** [supabase/supabase#2595] [Questions] [answered]
My employer gives me education days each year and, usually, I use that time to build open-source python stuff. This year I want to build a live service instead and supabase seems like a fun stack to build on top of. I'm more of a python machine learning person than a web developer though, so I figured I should double-check if my ideas surrounding security are alright. 

I want to make a service that allows a user to send themselves an email, with a delay of 30 days. I figured it'd be a cool way to run a diary of sorts. The main idea is that you don't need to log in, you merely need to send an email to a public email address. Here's the setup that I'm considering. 

1. I'll be using [postmark](https://postmarkapp.com/). When an email comes in postmark picks it up and will post it to a little service, probably a docker container, which takes care of writing into supabase. 
2. The container will also listen to a cron trigger on a daily basis to check the supabase table if there are emails that we can send back. These emails are forwarded to postmark and once postmark receives the email we can remove it from supabase. 

![image](https://user-images.githubusercontent.com/1019791/127185482-cc05f73c-b43a-4b5d-b0ec-3debbbf57206.png)

There are a few things that I am thinking about. 

0. The messages sent to my service might certainly be very personal, so I gotta take proper effort here to keep things safe. 
1. I think I should ~~hash~~ encrypt the email and message that I store in the supabase table. That way, should any nefarious person gain access to the table, no private information is shared. Do folks agree?
2. My container will need access to keys. I could use environment variables for this or perhaps do something with a `.env` file. I gotta be careful here though, if somebody gains access to my container they also have access to all the keys. Something is telling me that setting env variables from the cloud service that is hosting the container is safest. 

Do folks agree? Or are there some extra permission tricks I may do using supabase here?

**Comment by inian** [comment, marked as answer]
Hey Vincent, sounds like a fun project! Few comments
- you might be able t completely eliminate the backend service and do everything with just Supabase. You can use Postgres as a [cron server](https://supabase.io/blog/2021/03/05/postgres-as-a-cron-server) and you might be able to call Postmark via their API using via [pg_http](https://github.com/pramsey/pgsql-http) which is an extension installed in your Supabase database
- I think you probably want to encrypt your emails instead of hashing them since you can't get the original content back if you hash them. If you are using a separate backend service, the encryption key can be present in the container - so even if an attacker gains control over the database, the content is encrypted. 
- As for storing the secret safely, you store it in an .env file or dynamically pull in the encryption keys from a secrets manager (like Hashicorp Vault or AWS secrets manager) and bootstrap your application with that. That way the secret is not in your environment variable which other processes can access too (if you are worried about that) and only in your application memory.

**Reply by koaning** [reply]
Gah! Typo'd there. But yes, I meant encrypt, not hash. 

Thanks for the reply :)
